### ISSUE TITLE
Databricks: Cannot register sklearn model because of Metadata issue when using workspace experiment

### ISSUE THREAD

**Issue by claywey** [mlflow/mlflow#3490]
I was working with the Pycaret library to log models to mlflow. The models were logged to the a workspace experiment and the artifacts were save to an Azure datalake mount point. However, I received this warning. Note: I am running Databricks Runtime 7.3ML

"WARNING mlflow.models.model: Logging model metadata to the tracking server has failed, possibly due older server version. The model artifacts have been logged successfully under"

Everything appear fine. I can see all of my logged runs and I can see my sklearn artifacts from the workspace experiment UI. However, when I go to register the the model to the Model Registry I get this message:

"Failed registration. The given source path does not exist."

I thought maybe this was a Pycaret issue so I ran some sample code from the Mlflow documentation. The below code seems to work and logs the model under the Notebook experiment.

> import mlflow
> import mlflow.sklearn
> from sklearn.datasets import load_iris
> from sklearn import tree
> 
> iris = load_iris()
> sk_model = tree.DecisionTreeClassifier()
> sk_model = sk_model.fit(iris.data, iris.target)
> 
> mlflow.log_param("criterion", sk_model.criterion)
> mlflow.log_param("splitter", sk_model.splitter)
> 
> mlflow.sklearn.log_model(sk_model, "sk_models")

But, if I use mlflow.set_experiment() to log my parameters and model to a Workspace experiment I do not receive the earlier warning but I get the same error when attempting to register the model. 

I would like to be able to save my model artifacts to my mount point and be able to register them from that path. However, it appears I cannot do that.

**Comment by Mika24** [comment]
I have the similar issue. 
I log and store the artifacts (conda.yml, MLModel, mode file) in the ml flow tracking server in Azure Databricks. But afterward, I download these file and upload them to Azure DataLake storage. Then I tried to read from the DataLake storage with "load_model" and I can see the model is loaded successfully and the metadata of model is also available.

However, after I registered the model, although the operation seemed to succeed, in the "Model" UI in Databricks, it said the same error "Failed registration. The given source path does not exist.". Therefore, I can't serve the model or use it as an udf.

My goal is able to somehow make the ML models available in another Azure Databricks instance.

**Comment by amesar** [comment]
Probably because your new imported registered model is still pointing to the run URI on the old tracking server. You would need to import the run too. Migrating experiments/runs/models isn't well articulated in MLflow.
This may help: https://github.com/amesar/mlflow-tools/tree/master/mlflow_tools/export_import

**Comment by Mika24** [comment]
> Probably because your new imported registered model is still pointing to the run URI on the old tracking server. You would need to import the run too. Migrating experiments/runs/models isn't well articulated in MLflow.
> This may help: https://github.com/amesar/mlflow-tools/tree/master/mlflow_tools/export_import

Thanks Andre for the guidance. Will definitely try it out.

**Comment by dmatrix** [comment]
@Mika24 Let us know if this works, so we can close this issue.
cheers!

**Comment by claywey** [comment]
@dmatrix My original issue is a bit different from @Mika24 issue.

I am working within Databricks. When I have a workspace experiement which is logging and saving aritifcats to my Azure Datalake, I expect to register a model from that workspace experiment given that artifact location. However, Mlflow is not logging the metadata correctly. When I go to register the model from the UI, it says that "Failed registration. The given source path does not exist" despite Mlflow having saved the artificats to my datalake itself. The metadata tracking doesnt appear to log that the models are located on my datalake and not within the Databricks mlflow filestore

**Comment by dmatrix** [comment]
@claywey 
Can you try something simple on your Azure Databricks?

```
[def print_experiment_info(experiment):
  print("Name: {}".format(experiment.name))
  print("Experiment ID: {}".format(experiment.experiment_id))
  print("Artifact Location: {}".format(experiment.artifact_location))
  print("Lifecycle_stage: {}".format(experiment.lifecycle_stage))
  
def print_run_info(run):
  print("run_id: {}".format(run.info.run_id))
  print("experiment_id: {}".format(run.info.experiment_id))
  print("params: {}".format(run.data.params))
  print("artifact_uri: {}".format(run.info.artifact_uri))
  print("status: {}".format(run.info.status))

import mlflow
import mlflow.sklearn
import numpy as np

from sklearn.linear_model import LinearRegression

print(mlflow.__version__)

mlflow.sklearn.autolog()
# train a model
model = LinearRegression()

# Create four runs
for i in range(4):
  with mlflow.start_run() as run:
    X = np.array([[1, 1*i], [1, 2*i], [2, 2*i], [2, 3*i]])
    y = np.dot(X, np.array([1, 2])) + i
    model.fit(X, y)
    print("run_id: {}".format(run.info.run_id))

from mlflow.tracking import MlflowClient
experiment = MlflowClient().get_experiment(<expriment_id>)
run = MlflowClient().get_run(<run_id>)

# Show experiment and run info
print_experiment_info(experiment)
print_run_info(run)](url)

```

```
Output on Databricks:

Experiment info:
Name: /Users/<your_email>/artifacts_registry_locations
Experiment ID: 30306676
Artifact Location: dbfs:/databricks/mlflow-tracking/30306676
Lifecycle_stage: active

Run Info:
run_id: 57137e0d7f0d4cb0b20c857977cde3fb
experiment_id: 30306676
params: {'copy_X': 'True', 'fit_intercept': 'True', 'n_jobs': 'None', 'normalize': 'False'}
artifact_uri: dbfs:/databricks/mlflow-tracking/30306676/57137e0d7f0d4cb0b20c857977cde3fb/artifacts
status: FINISHED
```

Are the artifact URI from experiment and run the same? For Azure will be blog URI

**Comment by claywey** [comment]
Here's the output from the first one. It is logging to a notebook level experiement. I am able to register the model from the UI with this run.

> 1.11.0
run_id: 249ebbce54ab41d199fce8fc3ad31ad4
run_id: e26839f30b08403db237a6dac7deace2
run_id: 88fb4a17d04244c58db4f2229b9ca5b5
run_id: caaad72d41494117b9507bd81b20185c
Name: /Users/clayton.weymouth@lucinahealth.com/mlflowtest
Experiment ID: 1544858569275687
Artifact Location: dbfs:/databricks/mlflow-tracking/1544858569275687
Lifecycle_stage: active
run_id: caaad72d41494117b9507bd81b20185c
experiment_id: 1544858569275687
params: {'copy_X': 'True', 'fit_intercept': 'True', 'n_jobs': 'None', 'normalize': 'False'}
artifact_uri: dbfs:/databricks/mlflow-tracking/1544858569275687/caaad72d41494117b9507bd81b20185c/artifacts
status: FINISHED
registry_uri: databricks
tracking_uri: databricks

For the next one, I added mlflow.set_experiment to log the same thing but to my workspace level experiment.  My artifacts are saved to my lake. And, within the Mlflow UI, I can clearly see all artificats and their correct locations. However, I cannot register models because it of the source path not existing.

> 1.11.0
run_id: 7750ed8095ca45e2817c762aba11213e
run_id: 9aad48b50916402a9b6d29d415efb054
run_id: c7ebb42c9ca7447e8193443f0a189cf9
run_id: f67fed4f308a448bb1b531d6a0a97e40
Name: /Shared/lucina/BaselineTwo
Experiment ID: 1544858569275689
Artifact Location: dbfs:/mnt/lake/modeling/mlflow
Lifecycle_stage: active
run_id: f67fed4f308a448bb1b531d6a0a97e40
experiment_id: 1544858569275689
params: {'copy_X': 'True', 'fit_intercept': 'True', 'n_jobs': 'None', 'normalize': 'False'}
artifact_uri: dbfs:/mnt/lake/modeling/mlflow/f67fed4f308a448bb1b531d6a0a97e40/artifacts
status: FINISHED
registry_uri: databricks
tracking_uri: databricks

**Comment by dmatrix** [comment]
@claywey if you set and experiment then make sure that when you start that run, that run is logged under that experiment if that's where you want your runs to reside under.

For example,

```
from mlflow.tracking import MlflowClient
experiment_id = MlflowClient().create_experiment("/Users/jules@databricks.com/issue_3490")

 with mlflow.start_run(experiment_id=experiment_id):
     .....
     ...

```

**Comment by claywey** [comment]
@dmatrix 

The runs are being logged and I can see them clearly. In the attached photo, everything circled in red is as I expect. This run logged my run, created and stored my artifacts to my datalake. But if you look at the area circled in blue, that is where I am getting the "Failed registration. The given source path does not exist." error. I am not able to add the model to the model registry. This is one of the runs I created via the code you provided earlier. (Note: "iris" is just the test model name I have been using)


![Mlflow](https://user-images.githubusercontent.com/57914269/95639490-96ed2d00-0a66-11eb-9a1c-c5bb611db255.PNG)

**Comment by dmatrix** [comment]
@claywey Hmm, this is a bit weird. I would start anew. 

1. Delete the runs
2. Delete the experiment


Now try without creating or setting an experiment. Just the Default experiment.

Is `Lucina` your login id?  Are you setting any artifact location? What is the /mnt/lake?

**Comment by dmatrix** [comment]
> The default experiment is the experiment scoped to the Notebook itself. That works. But this is a huge problem because it is storing all of my aritifact on the Databricks VM. I would much prefer storing it on my lake. mnt/lake/ is the dbfs mount point for my DataLake Gen 1 Storage

Artifacts are persisted on a permanent DBFS location, and all MLflow run entities are persisted in MySQL. You can provide artifact_location to be any[ s3: URI.](https://docs.databricks.com/applications/mlflow/tracking.html#workspace-experiments) 

"Databricks supports DBFS, S3, and Azure Blob storage artifact locations.
To store artifacts in S3, specify a URI of the form s3://<bucket>/<path>. MLflow obtains credentials to access S3 from your clusters’s instance profile. Artifacts stored in S3 cannot be viewed in the MLflow UI; you must download them using an object storage client." 

>  I would much prefer storing it on my lake. mnt/lake/ is the dbfs mount point for my DataLake Gen 1 Storage

There might be a permission issue. Your scenario should work. Are you a databricks customer? If you file a ES-ticket and provide a reproducible step to create this problem.

**Comment by kennydataml** [comment]
@dmatrix this isn't a permission issue. this is an mlflow registry problem. I'm facing the same problem that you cannot register a model that is sitting in Azure Blob storage. It must sit within internal DBFS for you to register the model. pretty sure this is a limitation that needs to be resolved.

registering an artifact is different from registering a model...

**Comment by ru2nuts** [comment]
seeing this issue with Azure blob mounted under /dbfs/mnt/...
Confirming - this is not a permission issue, I can read/write from databricks notebooks, I **can log** parameters/metrics/models from an experiment run. The issue is only with **registering** the models. This issue is critical for ability to use MLflow on Azure.

**Comment by EMCP** [comment]
> Probably because your new imported registered model is still pointing to the run URI on the old tracking server. You would need to import the run too. Migrating experiments/runs/models isn't well articulated in MLflow.
> This may help: https://github.com/amesar/mlflow-tools/tree/master/mlflow_tools/export_import

This is now 404'd .. was googling around after seeing a similar issue with PyTorch models

**Comment by noderat** [comment]
I am also encountering the same problem when trying to register a model when the MLFlow experiment's artifact path is set to a mounted ADLSv2 path.

Changing a working notebook from `mlflow.create_experiment(name="test")` to `mlflow.create_experiment(name="test", artifact_location="dbfs:/mnt/adls_gen2_mount_example/")` fails when trying to register the model with the model registry both with the API and the MLFlow UI.

**Comment by SiebertLooijeAtBillyGrace** [comment]
I am facing the same issue. Did anyone has a solution for it?

**Comment by Sebastianjoergensen** [comment]
Thanks to the people spending time to answer this thread.

At the time of writing I can confirm that the MLFlow model registry does not work for mounted storage on Databricks in Azure Cloud. Only the default DBFS works with the MLFlow model registry. 

For legislative reasons we cannot use the default DBFS that is hosted on a Microsoft managed ressource group (using public IPs).

This is a dealbreaker that I would have like to have known much earlier. 

There are 2 problems:

1: The MLFLow model registry is incompatible with mounted storage on Azure Cloud
2: This is not clearly described in the documentation.


Speaking in euphemisms, both problems are a big source of frustration.

**Comment by SergioOlivero** [comment]
Hi! I'm having the exact same issue but with a mounted S3 bucket. Is it the same conclusion? Models can not be registered if the experiment is saved in any other place other than the default DBFS?

**Comment by Sebastianjoergensen** [comment]
Hello Sergio

I see it is now possible to register models in Unity Catalog. It seems it might be an actual solution to this Github issue.

Azure: https://learn.microsoft.com/en-us/azure/databricks/machine-learning/manage-model-lifecycle/
AWS: https://docs.databricks.com/en/machine-learning/manage-model-lifecycle/index.html

At the time of writing I have not yet tested it out myself.
